src/main/drivers/dshot_bitbang_impl.h

```diff
@@ -0,0 +1,228 @@
+
+#ifdef USE_HAL_DRIVER
+#define BB_GPIO_PULLDOWN GPIO_PULLDOWN
+#define BB_GPIO_PULLUP   GPIO_PULLUP
+#else
+#define BB_GPIO_PULLDOWN GPIO_PuPd_DOWN
+#define BB_GPIO_PULLUP   GPIO_PuPd_UP
+#endif
+
+#ifdef USE_DMA_REGISTER_CACHE
+typedef struct dmaRegCache_s {
+#if defined(STM32F4) || defined(STM32F7)
+    uint32_t CR;
+    uint32_t FCR;
+    uint32_t NDTR;
+    uint32_t PAR;
+    uint32_t M0AR;
+#else
+#error No MCU dependent code here
+#endif
+} dmaRegCache_t;
+#endif
+
+// Per pacer timer
+
+typedef struct bbPacer_s {
+    TIM_TypeDef *tim;
+    uint16_t dmaSources;
+} bbPacer_t;
+
+// Per GPIO port and timer channel
+
+typedef struct bbPort_s {
+    //int portIndex; // XXX Not used if managed by gpioToMotorPort[] array
+    GPIO_TypeDef *gpio;
+    const timerHardware_t *timhw;
+#ifdef USE_HAL_DRIVER
+    uint32_t llChannel;
+#endif
+
+    uint16_t dmaSource;
+
+    dmaResource_t *dmaResource; // DMA resource for this port & timer channel
+    uint32_t dmaChannel;        // DMA channel or peripheral request
+
+    uint8_t direction;
+
+#ifdef USE_DMA_REGISTER_CACHE 
+    // DMA resource register cache
+    dmaRegCache_t dmaRegOutput;
+    dmaRegCache_t dmaRegInput;
+#endif
+
+    // For direct manipulation of GPIO_MODER register
+    uint32_t gpioModeMask;
+    uint32_t gpioModeInput;
+    uint32_t gpioModeOutput;
+
+    // Idle value
+    uint32_t gpioIdleBSRR;
+
+#ifdef USE_HAL_DRIVER
+    LL_TIM_InitTypeDef timeBaseInit;
+#else
+    TIM_TimeBaseInitTypeDef timeBaseInit;
+#endif
+
+    // Output
+    uint16_t outputARR;
+#ifdef USE_HAL_DRIVER
+    LL_DMA_InitTypeDef outputDmaInit;
+#else
+    DMA_InitTypeDef outputDmaInit;
+#endif
+    uint32_t *portOutputBuffer;
+    uint32_t portOutputCount;
+
+    // Input
+    uint16_t inputARR;
+#ifdef USE_HAL_DRIVER
+    LL_DMA_InitTypeDef inputDmaInit;
+#else
+    DMA_InitTypeDef inputDmaInit;
+#endif
+    uint16_t *portInputBuffer;
+    uint32_t portInputCount;
+    bool inputActive;
+
+    // Misc
+#ifdef DEBUG_COUNT_INTERRUPT
+    uint32_t outputIrq;
+    uint32_t inputIrq;
+#endif
+} bbPort_t;
+
+// Per motor output
+
+typedef struct bbMotor_s {
+    dshotProtocolControl_t protocolControl;
+    int pinIndex;            // pinIndex of this motor output within a group that bbPort points to
+    IO_t io;                 // IO_t for this output
+    uint8_t output;
+    uint32_t iocfg;
+    bbPort_t *bbPort;
+    bool configured;
+    bool enabled;
+} bbMotor_t;
+
+#define MAX_MOTOR_PACERS  2
+extern FAST_RAM_ZERO_INIT bbPacer_t bbPacers[MAX_MOTOR_PACERS];  // TIM1 or TIM8
+extern FAST_RAM_ZERO_INIT int usedMotorPacers;
+
+extern FAST_RAM_ZERO_INIT bbPort_t bbPorts[MAX_SUPPORTED_MOTOR_PORTS];
+extern FAST_RAM_ZERO_INIT int usedMotorPorts;
+
+extern FAST_RAM_ZERO_INIT bbMotor_t bbMotors[MAX_SUPPORTED_MOTORS];
+
+extern FAST_RAM_ZERO_INIT bbPort_t *gpioToMotorPort[10]; // GPIO group to bbPort mapping
```

Comment:
Can we get rid of the magic number here?

But it would probably be cleaner to change this to only use `bbPorts`, and iterate through them to find the one with the right port index.

Reply:
done